libethereum/EthereumPeer.cpp

```diff
@@ -246,22 +246,134 @@ bool EthereumPeer::interpret(unsigned _id, RLP const& _r)
 		host()->onPeerTransactions(dynamic_pointer_cast<EthereumPeer>(dynamic_pointer_cast<EthereumPeer>(shared_from_this())), _r);
 		break;
 	}
-	case GetBlockHashesPacket:
+	case GetBlockHeadersPacket:
 	{
-		h256 later = _r[0].toHash<h256>();
-		unsigned limit = _r[1].toInt<unsigned>();
-		clog(NetMessageSummary) << "GetBlockHashes (" << limit << "entries," << later << ")";
-		unsigned c = min<unsigned>(host()->chain().number(later), min(limit, c_maxHashesToSend));
+		/// Packet layout:
+		/// [ block: { P , B_32 }, maxHeaders: P, skip: P, reverse: P in { 0 , 1 } ]
+		const auto blockId = _r[0];
+		const auto maxHeaders = _r[1].toInt<u256>();
+		const auto skip = _r[2].toInt<u256>();
+		const auto reverse = _r[3].toInt<bool>();
+
+		auto& bc = host()->chain();
+
+		auto numHeadersToSend = maxHeaders <= c_maxHeadersToSend ? static_cast<unsigned>(maxHeaders) : c_maxHeadersToSend;
+
+		if (skip > std::numeric_limits<unsigned>::max() - 1)
+		{
+			clog(NetAllDetail) << "Requested block skip is too big: " << skip;
+			break;
+		}
+
+		auto step = static_cast<unsigned>(skip) + 1;
+		assert(step > 0 && "step must not be 0");
```

Comment:
What i mean is that `step` can't be 0 here since it is already checked by if clause above